Upgrade EpiVis dashboard

rzats opened the pull request:
Fixes:
- https://cmu-delphi.atlassian.net/browse/OKRS24-61
- https://cmu-delphi.atlassian.net/browse/OKRS24-62
- https://cmu-delphi.atlassian.net/browse/OKRS24-63

melange396 commented on src/components/Chart.svelte:
```diff
@@ -380,16 +381,46 @@
-  export function fitData(shouldAnimate = false): void {
-    if (datasets.length === 0) {
+  // Fit viewport to the currently available datasets.
+  // As the store can take a while to update, the arguments here include extra datasets
+  // to be included (in case a new one was just selected) or excluded (de-selected)
```
I am not sure what you mean here...  Which store is updating?  If there is some delay or lag being inserted somewhere, should we be using a callback to trigger this after said update has completed?

melange396 commented on src/tour.ts:
```diff
@@ -16,6 +16,15 @@ export const tour = new Shepherd.Tour({
 const cancel = tour.cancel.bind(tour);
 
+// Dismiss tour on future runs once it is cancelled or completed.
+function dismissTour() {
+  if (!localStorage.getItem('shepherd-tour')) {
+    localStorage.setItem('shepherd-tour', 'yes');
+  }
+}
+tour.on('cancel', dismissTour);
+tour.on('complete', dismissTour);
```
There is useful information in this tour, we should not automatically make it forever inaccessible to the user...  Can we change "cancel" to "dismiss" and add a checkbox for "don't show this on startup", and then create a "question mark" button in the top row of the application to re-show the dialog?

melange396 reviewed (changes requested):
Do you have a preview instance/installation for this branch somewhere?

I think we need make the automatic `fitData()` calls a toggle-able option, perhaps its own "mode" (joining the three existing modes already available in the top menu).  For instance, if a user is looking at a number of signals in a particular time window, and they turn some off and on to get a better look at the other signals, they will lose the specific view they had.

One way to handle this would be that we could switch to "auto-fit mode" when adding a new signal to the canvas (but not when simply toggling an existing signal).  While in "auto-fit mode", clicking the mouse inside the view canvas or using the scroll wheel should switch to "pan mode".  It might be nice to be able to opt out of "auto-fit mode" when adding new signals, which could be done with a checkbox in the signal load dialog.

rzats commented:
Notes from the meeting:

- Reword the comment about `include`/`exclude` to make sure it refers specifically to the `dataset` variable
- Add a button to disable the new autofitting (but it's enabled by default)

rzats commented:
@melange396 updated this PR as well!

dshemetov reviewed (changes requested):
Some notes:

* [x] https://cmu-delphi.atlassian.net/browse/OKRS24-62 looks done, the tour doesn't show up every time and now there's a button to show it later
  * [x] However the 'h' hotkey doesn't bring up the help tour (other keyboard commands work for me)
* [ ] When I click import datasets from API and I select FluView, the app hangs indefinitely, which didn't happen before. Something about the way the initial dataset is loaded might be off?
* [ ] https://cmu-delphi.atlassian.net/browse/OKRS24-61 the sample signals is removed, but I'm not sure what "match defaults with COVIDcast dashboard" means; do we want FluView to be the default? if so, this is done, if not, then we should decide what to display
* [x] https://cmu-delphi.atlassian.net/browse/OKRS24-63 zoom to signals seems to work well, I've tried importing a number of signals and sources, turning them on and off works quite seamlessly
* [ ] I see the EpiVis sample is still in the menu, though deactivated. We should probably remove this completely, seems unnecessary and distracting.
* [ ] I would choose a different icon than a gear for the autofit functionality. Gear made me think it was a "general settings" button. Maybe something more like [this](https://fontawesome.com/icons/think-peaks?f=brands&s=solid) or [like this](https://fontawesome.com/icons/up-down?f=classic&s=solid)?

rzats commented:
@dshemetov this is ready for a re-review! A couple of notes:

> the 'h' hotkey doesn't bring up the help tour

Fixed!

> When I click import datasets from API and I select FluView, the app hangs indefinitely, which didn't happen before. Something about the way the initial dataset is loaded might be off?

This would happen before if you were to try loading the same dataset twice. I added an extra check for that.

> I'm not sure what "match defaults with COVIDcast dashboard" means; do we want FluView to be the default? if so, this is done, if not, then we should decide what to display

IMO this is done, but I can set the default to something else as well.

> I see the EpiVis sample is still in the menu, though deactivated. We should probably remove this completely, seems unnecessary and distracting.

Done!

> I would choose a different icon than a gear for the autofit functionality. Gear made me think it was a "general settings" button. Maybe something more like [this](https://fontawesome.com/icons/think-peaks?f=brands&s=solid) or [like this](https://fontawesome.com/icons/up-down?f=classic&s=solid)?

Done, went for the `up-down` icon.

dshemetov reviewed (changes requested):
My previous concerns have been addressed, thanks! 

If we want to be strict with the first JIRA ticket: 
- [ ] "match defaults with COVIDcast dashboard" - the default signal for COVIDcast dashboard is COVID Hospitalizations, so that would be better
  - [ ] a subissue here is that I can't seem to select US National when using the COVID Hospitalization dataset; is this a bug or an api limitation or something else?
- [ ] I'm getting very strange behavior with automatic resizing when adding the COVID Hospitalizations as a dataset; when the dropdown for the COVID Hosp signals is closed, the sizing is right, but when the dropdown is open, the sizing gets distorted, the plot stretches down (see the images below). It's like it considers the bottom of the scrollbar on the left to be the bottom of the viewport on the right?

<img width="1171" alt="image" src="https://github.com/cmu-delphi/www-epivis/assets/1810426/886c30ed-c12c-42cc-96b4-7ef1808984b4">
<img width="1337" alt="image" src="https://github.com/cmu-delphi/www-epivis/assets/1810426/f973548e-7d29-4880-b4e0-4613d918ca43">

rzats commented:
@dshemetov 

> a subissue here is that I can't seem to select US National when using the COVID Hospitalization dataset; is this a bug or an api limitation or something else?

It is indeed an API limitation! That dataset works on the [state level only](https://cmu-delphi.github.io/delphi-epidata/api/covid_hosp.html). It seems to be a subset of the [overall HHS dataset](https://cmu-delphi.github.io/delphi-epidata/api/covidcast-signals/hhs.html), which does offer more geo resolutions.

> very strange behavior with automatic resizing when adding the COVID Hospitalizations as a dataset

This seems to be a CSS issue: if too many datasets are added to the left sidebar, it overflows offscreen. I've fixed this with the latest commit!

melange396 reviewed (changes requested):
This looks great, but there are things id like to look at before we merge.

- I made a number of suggestions for verbiage on the "tour", a few readability suggestions, and one for inverting the color on toggled buttons.

- I see what youre doing with the `include` and `exclude` arguments to `Chart.svelte:fitData()`, but something tells me those shouldnt be necessary...  I thought i figured out some ways around it and put them into #41, but now i cant get that to generate a Netlify preview to test.  Do you know what i am missing?  Do you think that approach has merit?

- This current implementation doesnt do the automatic "fit" operation on page load when a signal is specified in the base64-encoded url (such as [this one](https://deploy-preview-36--cmu-delphi-epivis.netlify.app/#eyJkYXRhc2V0cyI6W3siY29sb3IiOiIjNDE1NzQyIiwidGl0bGUiOiJ2YWx1ZSIsInBhcmFtcyI6eyJfZW5kcG9pbnQiOiJjb3ZpZGNhc3QiLCJkYXRhX3NvdXJjZSI6ImRvY3Rvci12aXNpdHMiLCJzaWduYWwiOiJzbW9vdGhlZF9jbGkiLCJ0aW1lX3R5cGUiOiJkYXkiLCJnZW9fdHlwZSI6InN0YXRlIiwiZ2VvX3ZhbHVlIjoiYXIifX1dfQ==); it requires a post-load press of the "fit data to screen" button before the signal is visible).  I think my exploration in #41 might  address this (or open a easier avenue to accomplish it), but as i said, its not currently rendering a preview.

- obviously not a result of the changes in this PR, but worth mentioning: there is a typo which alters the tooltip behavior for one of the buttons, please `s/tootlip/tooltip/` on https://github.com/cmu-delphi/www-epivis/blob/d48e3b03812ab7704d10fb367d179787a9eb7f2d/src/components/TopMenu.svelte#L134

And then there are some things that dont need to be in this PR, but i will make separate issues for them:

- support `time_type:week` (aka EpiWeek format time values) for covidcast signals (first priority is to support them in signals [encoded in url param](https://github.com/cmu-delphi/www-epivis/blob/6e310a147578038a8bad94ff9525505afbff465b/src/deriveLinkDefaults.ts#L167-L172), then to support them in the signal picker) -- see https://github.com/cmu-delphi/www-epivis/blob/6e310a147578038a8bad94ff9525505afbff465b/src/api/EpiData.ts#L181-L182 
- turn autofit into a mode, as mentioned in https://github.com/cmu-delphi/www-epivis/pull/36#pullrequestreview-1945159659

rzats commented on src/components/tree/TreeInnerNode.svelte:
```diff
@@ -28,7 +30,7 @@
     {#each node.datasets as child (child.title)}
       {#if child instanceof DataSet}
-        <TreeLeafNode node={child} />
+        <TreeLeafNode {chart} node={child} />
```
The Svelte linter actually reverts this change when I try to apply it :) I could disable this in the config, but shorthand attributes like this show up a lot in the code base, e.g. a couple lines before:
```
<side class="left" {style} data-tour="browser">
                   ^^^^^^^
```
and elsewhere:
```
<select class="uk-select" required {name} {id} bind:value>
                                   ^^^^^^ ^^^^
```
So I think it might be fairly clear what's going on with these.

melange396 replied:
i dont love it, but if thats the preferred style in Svelte, we might as well go with it.

rzats commented on src/components/Chart.svelte:
```diff
@@ -1,6 +1,7 @@
 <script lang="ts">
   import { onMount } from 'svelte';
   import type DataSet from '../data/DataSet';
+  import type DataGroup from '../data/DataSet';
```
We no longer `import type DataGroup` here after #41, so doesn't seem to be needed anyhow.

rzats commented:
@melange396 applied all of your suggestions except a particular codestyle one!

> `include` and `exclude` arguments in `Chart.svelte:fitData()`
These are removed in #41...

> no automatic "fit" when a signal is specified in the base64-encoded URL
...and I doublechecked that this is now working as well.

melange396 commented on src/components/tree/TreeLeafNode.svelte:
```diff
@@ -13,6 +15,12 @@
   }
+  $: {
+    // runs whenever $activeDatasets is updated
+    if ($activeDatasets && chart && $autoFit === true) {
```
```suggestion
    if ($activeDatasets && chart && $autoFit) {
```

melange396 commented on src/components/TopMenu.svelte:
```diff
@@ -92,8 +102,22 @@
-      class="uk-button uk-button-default uk-button-small"
+      class="uk-button uk-button-small"
+      disabled={!chart}
+      class:uk-active={$autoFit === true}
+      class:uk-button-secondary={$autoFit === true}
+      class:uk-button-default={$autoFit === false}
```
```suggestion
      class:uk-active={$autoFit}
      class:uk-button-secondary={$autoFit}
      class:uk-button-default={! $autoFit}
```

melange396 commented on src/components/TopMenu.svelte:
```diff
@@ -92,8 +102,22 @@
+      type="button"
+      class="uk-button uk-button-small"
       class:uk-active={$isShowingPoints}
+      class:uk-button-secondary={$isShowingPoints === true}
+      class:uk-button-default={$isShowingPoints === false}
```
```suggestion
      class:uk-button-secondary={$isShowingPoints}
      class:uk-button-default={! $isShowingPoints}
```

melange396 reviewed (changes requested):
a few boolean comparison nits, but otherwise this looks fantastic and i think its ready to go!

melange396 commented:
You should probably also bump the version number for this: https://github.com/cmu-delphi/www-epivis/blob/6e310a147578038a8bad94ff9525505afbff465b/package.json#L3

rzats commented:
@melange396 applied the changes!

melange396 reviewed (approved):
nice!
